apps/prairielearn/src/lib/cookie.ts

```diff
@@ -10,3 +10,157 @@ export function shouldSecureCookie(req: Request): boolean {
+      httpOnly: true,
+    },
+  },
+  {
+    oldName: 'preAuthUrl',
+    newName: 'pl2_pre_auth_url',
+  },
+  {
+    oldName: 'pl_pw_origUrl',
+    newName: 'pl2_pw_original_url',
+  },
+  {
+    oldName: 'pl_disable_auto_authn',
+    newName: 'pl2_disable_auto_authn',
+  },
+  {
+    oldName: 'pl_requested_data_changed',
+    newName: 'pl2_requested_data_changed',
+  },
+  {
+    oldName: 'pl_requested_uid',
+    newName: 'pl2_requested_uid',
+  },
+  {
+    oldName: 'pl_requested_course_role',
+    newName: 'pl2_requested_course_role',
+  },
+  {
+    oldName: 'pl_requested_course_instance_role',
+    newName: 'pl2_requested_course_instance_role',
+  },
+  {
+    oldName: 'pl_requested_mode',
+    newName: 'pl2_requested_mode',
+  },
+  {
+    oldName: 'pl_requested_date',
+    newName: 'pl2_requested_date',
+  },
+  {
+    oldRegexp: /pl_authz_workspace_(\d+)/,
+    oldPattern: 'pl_authz_workspace_$1',
+    newRegexp: /pl2_authz_workspace_(\d+)/,
+    newPattern: 'pl2_authz_workspace_$1',
+    options: {
+      maxAge: config.workspaceAuthzCookieMaxAgeMilliseconds,
+      httpOnly: true,
+    },
+  },
+];
+
+export function makeCookieMigrationMiddleware(
+  rewriteCookies: boolean,
+): (req: Request, res: Response, next: NextFunction) => void {
+  return function migrateCookiesIfNeededMiddleware(
+    req: Request,
+    res: Response,
+    next: NextFunction,
+  ) {
+    let didMigrate = false;
+
+    COOKIES_TO_MIGRATE.forEach((cookieMigration) => {
+      if (cookieMigration.oldName) {
+        // If the cookie was already migrated, we don't need to write anything
+        // back to the client. However, we'll overwrite the value of the old
+        // cookie in the current request so that application code is
+        // forward-compatible with the new cookie name.
+        if (req.cookies[cookieMigration.newName]) {
+          req.cookies[cookieMigration.oldName] = req.cookies[cookieMigration.newName];
+          return;
+        }
+
+        // If the old cookie isn't present, there's nothing to do.
+        if (!(cookieMigration.oldName in req.cookies)) return;
+
+        if (rewriteCookies) {
+          didMigrate = true;
+          res.cookie(cookieMigration.newName, req.cookies[cookieMigration.oldName], {
+            ...(cookieMigration.options ?? {}),
+            secure: shouldSecureCookie(req),
+          });
+        }
+      } else if (cookieMigration.oldRegexp) {
+        const { oldRegexp, oldPattern, newRegexp, newPattern } = cookieMigration;
+
+        // First, if needed, rewrite old cookies to use the new names.
+        const oldNames = req.cookies
+          ? Object.keys(req.cookies).filter((cookieName) => oldRegexp.test(cookieName))
+          : [];
+        if (oldNames.length) {
+          oldNames.forEach((oldName) => {
+            const newName = oldName.replace(oldRegexp, newPattern);
+
+            // If the cookie was already migrated, do nothing.
+            if (req.cookies[newName]) return;
```

Comment:
Do we want to set `oldName` to have the same value as `newName` here, to match the behavior above when we have an explicit `oldName`?

Reply:
We actually do this already, just in a separate loop below. This loop is handling old cookies; the latter is handling new cookies.